### PR TITLE
[Keyboard] add dual-direction diode capability for hbcp

### PR DESCRIPTION
**Pull request by hineybush**
## Description

So my PCB manufacturer messed up. Instead of common cathode diodes like I specced (MMBD4148CC) they used dual-direction (MMBD4148SE). This posed an issue as my original COL2ROW-only firmware would not work, understandably. @nooges from Keeb.io suggested that I look at PR #8160 and try these adjustments out. I built based on that PR and the PCB with the wrong diodes worked perfectly. I added the changes to an hbcp-specific matrix.c and adjusted the configs with @drashna's help. 

I'm adding the changes here because 1. it won't affect the "correct" HBCP PCBs with common cathode diode packages and 2. anyone building their own firmware from here, or updated VIA builds, won't be done by me - so the changes should be here for anyone to access. Once #8160 is merged, I will adjust the code within hbcp accordingly.

MMBD4148CC/SE datasheet here: https://www.onsemi.com/pub/Collateral/MMBD4148SE-D.pdf

## Types of Changes

- [ ] Core
- [x] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/hineybush/hbcp/config.h [hunk 1]
```diff
@@ -27,6 +27,9 @@ along with this program.  If not, see <http://www.gnu.org/licenses/>.
 #define PRODUCT         hbcp
 #define DESCRIPTION     compact 1800 layout custom keyboard
 
+/* needed because PCBA used the wrong diode part */
+#define EITHERWAY 3 /* Compensation for diode direction dissonance */
+
```
Comment (drashna) [inline keyboards/hineybush/hbcp/config.h:32]: Technically, this isn't needed right now, as the custom matrix doesn't check for it.

Comment (hineybush) [reply keyboards/hineybush/hbcp/config.h:32]: should I keep it for now, so it'll work with future master-merged code?

Comment (zvecr) [reply keyboards/hineybush/hbcp/config.h:32]: DIODE_DIRECTION would be the important bit (plus having this here before names and values are decided will potentially cause additional issues)
```suggestion
```

File: keyboards/hineybush/hbcp/rules.mk [hunk 2]
```diff
@@ -31,3 +31,5 @@ BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
 AUDIO_ENABLE = no           # Audio output on port C6
 FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
 HD44780_ENABLE = no 		# Enable support for HD44780 based LCDs
+CUSTOM_MATRIX = yes
```
Comment (zvecr) [inline keyboards/hineybush/hbcp/rules.mk:34]:
```suggestion
CUSTOM_MATRIX = lite
```

File: keyboards/hineybush/hbcp/matrix.c [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+#include "debounce.h"
+#include "quantum.h"
+
+static const pin_t row_pins[] = MATRIX_ROW_PINS;
+static const pin_t col_pins[] = MATRIX_COL_PINS;
+
+/* matrix state(1:on, 0:off) */
+static matrix_row_t raw_matrix[MATRIX_ROWS];  // raw values
+static matrix_row_t last_matrix[MATRIX_ROWS];  // raw values of last scan
+static matrix_row_t matrix[MATRIX_ROWS];  // debounced values
```
Comment (zvecr) [inline keyboards/hineybush/hbcp/matrix.c:30]:
```suggestion
static matrix_row_t last_matrix[MATRIX_ROWS]; // raw values of last scan
```

File: keyboards/hineybush/hbcp/matrix.c [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+__attribute__((weak)) void matrix_io_delay(void) { wait_us(30); }
+
+__attribute__((weak)) void matrix_init_kb(void) { matrix_init_user(); }
+
+__attribute__((weak)) void matrix_scan_kb(void) { matrix_scan_user(); }
+
+__attribute__((weak)) void matrix_init_user(void) {}
+
+__attribute__((weak)) void matrix_scan_user(void) {}
+
```
Comment (zvecr) [inline keyboards/hineybush/hbcp/matrix.c:42]:
```suggestion
```

File: keyboards/hineybush/hbcp/matrix.c [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    debounce(raw_matrix, matrix, MATRIX_ROWS, changed);
+
+    matrix_scan_quantum();
+    return (uint8_t)changed;
+}
+
+inline matrix_row_t matrix_get_row(uint8_t row) { return matrix[row]; }
+
+void matrix_print(void) {}
```
Comment (zvecr) [inline keyboards/hineybush/hbcp/matrix.c:170]:
```suggestion
```

File: keyboards/hineybush/hbcp/matrix.c [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        if (last_matrix[current_row] != raw_matrix[current_row]) {
+            changed = true;
+        }
+    }
+
+    debounce(raw_matrix, matrix, MATRIX_ROWS, changed);
+
+    matrix_scan_quantum();
+    return (uint8_t)changed;
+}
```
Comment (zvecr) [inline keyboards/hineybush/hbcp/matrix.c:166]:
```suggestion
bool matrix_scan_custom(matrix_row_t current_matrix[]) {
    bool changed = false;

    for (uint8_t current_row = 0; current_row < MATRIX_ROWS; current_row++) {
        last_matrix[current_row] = current_matrix[current_row];
        read_cols_on_row(current_matrix, current_row);
    }
    for (uint8_t current_col = 0; current_col < MATRIX_COLS; current_col++) {
        read_rows_on_col(current_matrix, current_col);
    }
    for (uint8_t current_row = 0; current_row < MATRIX_ROWS; current_row++) {
        if (last_matrix[current_row] != current_matrix[current_row]) {
            changed = true;
        }
    }

    return (uint8_t)changed;
}
```

File: keyboards/hineybush/hbcp/matrix.c [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+    // initialize matrix state: all keys off
+    for (uint8_t i = 0; i < MATRIX_ROWS; i++) {
+        raw_matrix[i] = 0;
+        matrix[i]     = 0;
+    }
+
+    debounce_init(MATRIX_ROWS);
+
+    matrix_init_quantum();
+}
```
Comment (zvecr) [inline keyboards/hineybush/hbcp/matrix.c:144]:
```suggestion
void matrix_init_custom(void) {
    // initialize key pins
    init_pins();
}
```

**Comment by zvecr** [comment]
>  it won't affect the "correct" HBCP PCBs with common cathode diode packages

As a passing comment, technically it will change scan speed which for some users might be less than ideal.

**Comment by drashna** [comment]
Ah, I see that @zvecr helped you get the lite custom matrix stuff implemented here. Good! :)

**Pull request merged**